### ISSUE TITLE
Align GCC family cross compilation across x86/arm (Bugzilla Bug 2004)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#9759]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 2004
Date: 2019-07-19T10:42:29+00:00
From: @leiflindholm
To:   @BobCF
CC:   f4bug, @cetola

Last updated: 2021-07-27T05:05:35+00:00

**Comment by tianocore-issues** [comment]
## Comment 8521
Date: 2019-07-19 10:42:29 +0000
From: @leiflindholm

- Industry Specification: ---
- Target OS: ---
- Bugzilla Assignee(s): @BobCF

Cross compilation has historically been (read: is currently being) done differently between x86 and arm.

ARM/AARCH64 profiles tend to use &lt;TOOLCHAIN&gt;_&lt;ARCH&gt;_PREFIX, leading to the form arm-linux-gnueabihf- or aarch64-linux-gnu, whereas IA32/X64 tend to use &lt;TOOLCHAIN&gt;_BIN, originally expecting to hold a path to a directory.
Adding to the confusion is clang, which is called with the same name for all supported architectures, but for ARM/AARCH64 uses &lt;TOOLCHAIN&gt;_&lt;ARCH&gt;_PREFIX for accessing the appropriate objcopy.

Meanwhile, the default values for CLANG38_IA32_TARGET/CLANG38_X64_TARGET means a non-native clang will not pick up the correct ld binary.

So, to summarise - it should be possible to orthogonally cross compile x86&lt;-&gt;arm using the same methods and commands.

**Comment by tianocore-issues** [comment]
## Comment 8541
Date: 2019-07-22 16:59:43 +0000
From: @lersek

The QEMU project&apos;s build scripts for edk2 (= for the OVMF and ArmVirtQemu fw platforms that are bundled with QEMU in binary form) honor this idiosyncrasy.

See file &quot;roms/edk2-funcs.sh&quot;, function &quot;qemu_edk2_get_cross_prefix_var&quot;.

Subscribing to this BZ as a reminder for Phil and myself -- whenever this BZ is fixed, and included in an edk2-stable release, the next edk2 submodule rebase in QEMU will have to update the build scripts in QEMU accordingly.

**Comment by tianocore-issues** [comment]
## Comment 8580
Date: 2019-07-25 20:20:14 +0000
From: @cetola

Bug Triage: Bob, please collect information for this issue.
